File: src/main/java/com/github/caijh/commons/util/Builder.java
package com.github.caijh.commons.util;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Consumer;
import java.util.function.Supplier;

public class Builder<T> {

    private final Supplier<T> supplier;
    private final List<Consumer<T>> modifiers = new ArrayList<>();

    private Builder(Supplier<T> supplier) {
        this.supplier = supplier;
    }

    public static <T> Builder<T> of(Supplier<T> supplier) {
        return new Builder<>(supplier);
    }

    public <V> Builder<T> with(BuilderConsumer<T, V> consumer, V v) {
        Consumer<T> c = t -> consumer.accept(t, v);
        modifiers.add(c);
        return this;
    }

    public T build() {
        T value = supplier.get();
        modifiers.forEach(modifier -> modifier.accept(value));
        modifiers.clear();
        return value;
    }

    @FunctionalInterface
    public interface BuilderConsumer<T, V> {

        void accept(T t, V v);

    }

}

File: src/main/kotlin/com/github/caijh/commons/util/Builder.kt
package com.github.caijh.commons.util

import java.util.function.Consumer
import java.util.function.Supplier

class Builder<T> private constructor(private val supplier: Supplier<T>) {

    private val modifiers: MutableList<Consumer<T>> = ArrayList()

    fun <V> with(consumer: BuilderConsumer<T, V>, v: V): Builder<T> {
        val c = Consumer { t: T -> consumer.accept(t, v) }
        modifiers.add(c)
        return this
    }

    fun build(): T {
        val value = supplier.get()
        modifiers.forEach(Consumer { modifier: Consumer<T> -> modifier.accept(value) })
        modifiers.clear()
        return value
    }

    fun interface BuilderConsumer<T, V> {
        fun accept(t: T, v: V)
    }

    companion object {
        @JvmStatic
        fun <T> use(supplier: Supplier<T>): Builder<T> {
            return Builder(supplier)
        }
    }
}
